pkg/dockerregistry/server/blobdescriptorservice.go

```diff
@@ -25,6 +38,137 @@ type blobDescriptorService struct {
 	distribution.BlobDescriptorService
 }
 
+// Stat returns a a blob descriptor if the given blob is either linked in repository or is referenced in
+// corresponding image stream. This method is invoked from inside of upstream's linkedBlobStore. It expects
+// a proper repository object to be set on given context by upper openshift middleware wrappers.
 func (bs *blobDescriptorService) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
-	return dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	// if there is a repo layer link, return its desc
+	desc, err := bs.BlobDescriptorService.Stat(ctx, dgst)
+	if err == nil {
+		return desc, err
+	}
+
+	// verify the blob is stored locally
+	desc, err = dockerRegistry.BlobStatter().Stat(ctx, dgst)
+	if err != nil {
+		return desc, err
+	}
+
+	// verify that an image holding this blob is actually present in the image stream
+	repo, found := RepositoryFrom(ctx)
+	if !found {
+		context.GetLogger(ctx).Errorf("failed to retrieve repository from context")
+		return distribution.Descriptor{}, distribution.ErrBlobUnknown
+	}
+
+	if imageStreamHasBlob(repo, dgst) {
+		return desc, err
+	}
+
+	return distribution.Descriptor{}, distribution.ErrBlobUnknown
+}
+
+// imageStreamHasBlob returns true if the given blob digest is referenced in image stream corresponding to
+// given repository. If not found locally, image stream's images will be iterated and fetched from newest to
+// oldest until found. Each processed image will update local cache of blobs.
+func imageStreamHasBlob(r *repository, dgst digest.Digest) bool {
+	repositories := r.cachedLayers.RepositoriesForDigest(dgst)
+	match := imageapi.DockerImageReference{Namespace: r.namespace, Name: r.name}.Exact()
+	for _, repo := range repositories {
```

Comment:
What's the expiration on the cache?  This check will allow someone to maintain access to an image for that period of time after their image is untagged.  In practice if the expiration is low we are ok, but if it is high admins may be at risk.  In practice a user may have a running pod already with the image, which means they can already keep access to an image after removed.

We *may* need to make the cache duration configurable and document the trade off.  At worst, the cache expiration needs to be in the low minutes (5-10).